examples/bsc-p2p/src/block_import/service.rs

```diff
@@ -0,0 +1,143 @@
+
+pub(crate) struct ImportService<T: EngineTypes> {
+    /// The handle to communicate with the engine service
+    engine: BeaconConsensusEngineHandle<T>,
+    /// Receive the new block from the network
+    from_network: UnboundedReceiver<IncomingBlock<T>>,
+    /// Send the outcome of the import to the network
+    to_network: UnboundedSender<Outcome<T>>,
+    /// Pending block imports.
+    pending_imports: FuturesUnordered<ImportFuture>,
+}
+
+impl<T: EngineTypes> ImportService<T> {
+    /// Create a new block import service
+    pub(crate) fn new(engine: BeaconConsensusEngineHandle<T>) -> (Self, ImportHandle<T>) {
+        let (handle, from_network, to_network) = ImportHandle::new();
+        (
+            Self { engine, from_network, to_network, pending_imports: FuturesUnordered::new() },
+            handle,
+        )
+    }
+
+    /// Add a new block import task to the pending imports
+    fn import(&mut self, block: BlockMsg<T>, peer_id: PeerId) {
+        let engine = self.engine.clone();
+        let to_network = self.to_network.clone();
+
+        let fut = async move {
+            let sealed_block = block.block.block.clone().seal();
+            let payload = T::block_to_payload(sealed_block);
+            let payload_result = engine.new_payload(payload).await;
+            let outcome = match payload_result {
+                Ok(payload_status) => match payload_status.status {
+                    PayloadStatusEnum::Valid => Outcome::<T> {
+                        peer: peer_id,
+                        result: Ok(BlockValidation::ValidBlock { block }),
+                    },
+                    PayloadStatusEnum::Invalid { validation_error } => Outcome::<T> {
+                        peer: peer_id,
+                        result: Err(BlockImportError::InvalidPayload(validation_error)),
+                    },
+                    _ => return,
+                },
+                Err(err) => {
+                    Outcome::<T> { peer: peer_id, result: Err(BlockImportError::Engine(err)) }
+                }
+            };
+
+            // TODO: FCU handling and send to engine
+
+            let _ = to_network.send(outcome);
+        };
+
+        self.pending_imports.push(Box::pin(fut));
+    }
+}
+
+impl<T: EngineTypes> Future for ImportService<T> {
+    type Output = Result<(), Box<dyn std::error::Error>>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // Process completed imports
+        let _ = self.pending_imports.poll_next_unpin(cx);
+
+        // Process new blocks
+        while let Poll::Ready(Some((block, peer_id))) = self.from_network.poll_recv(cx) {
+            self.import(block, peer_id);
+        }
+
+        // Check if channel is closed
+        if self.from_network.is_closed() {
+            error!("BSC block import service channel closed, service should terminate");
+            return Poll::Ready(Ok(()));
+        }
```

Comment:
we can ignore this, if we `fuse` the receiverstream, with the expectation that this channel should never be closed